services/wallet/collectibles/manager.go

```diff
@@ -254,68 +217,91 @@ func (o *Manager) FetchBalancesByOwnerAndContractAddress(ctx context.Context, ch
 	defer o.checkConnectionStatus(chainID)
 
-	anyProviderAvailable := false
+	cmd := circuitbreaker.Command{}
 	for _, provider := range o.accountOwnershipProviders {
 		if !provider.IsChainSupported(chainID) {
 			continue
 		}
-		anyProviderAvailable = true
 		if providerID != thirdparty.FetchFromAnyProvider && providerID != provider.ID() {
 			continue
 		}
 
-		assetContainer, err := provider.FetchAllAssetsByOwnerAndContractAddress(ctx, chainID, owner, contractAddresses, cursor, limit)
-		if err != nil {
-			log.Error("FetchAllAssetsByOwnerAndContractAddress failed for", "provider", provider.ID(), "chainID", chainID, "err", err)
-			continue
-		}
+		provider := provider
+		f := circuitbreaker.NewFunctor(
+			func() ([]interface{}, error) {
+				assetContainer, err := provider.FetchAllAssetsByOwnerAndContractAddress(ctx, chainID, owner, contractAddresses, cursor, limit)
+				if err != nil {
+					log.Error("FetchAllAssetsByOwnerAndContractAddress failed for", "provider", provider.ID(), "chainID", chainID, "err", err)
+				}
+				return []interface{}{assetContainer}, err
+			},
+		)
+		cmd.Add(f)
+	}
 
-		_, err = o.processFullCollectibleData(ctx, assetContainer.Items, true)
-		if err != nil {
-			return nil, err
-		}
+	if cmd.IsEmpty() {
+		return nil, ErrNoProvidersAvailableForChainID
+	}
 
-		return assetContainer, nil
+	cmdRes := o.getCircuitBreaker(chainID).Execute(cmd)
+	if cmdRes.Error() != nil {
+		log.Error("FetchAllAssetsByOwnerAndContractAddress failed for", "chainID", chainID, "err", cmdRes.Error())
+		return nil, cmdRes.Error()
 	}
 
-	if anyProviderAvailable {
-		return nil, ErrAllProvidersFailedForChainID
+	assetContainer := cmdRes.Result()[0].(*thirdparty.FullCollectibleDataContainer)
+	_, err := o.processFullCollectibleData(ctx, assetContainer.Items, true)
+	if err != nil {
+		return nil, err
 	}
-	return nil, ErrNoProvidersAvailableForChainID
+
+	return assetContainer, nil
 }
 
 func (o *Manager) FetchAllAssetsByOwner(ctx context.Context, chainID walletCommon.ChainID, owner common.Address, cursor string, limit int, providerID string) (*thirdparty.FullCollectibleDataContainer, error) {
 	defer o.checkConnectionStatus(chainID)
 
-	anyProviderAvailable := false
+	cmd := circuitbreaker.Command{}
 	for _, provider := range o.accountOwnershipProviders {
 		if !provider.IsChainSupported(chainID) {
 			continue
 		}
-		anyProviderAvailable = true
 		if providerID != thirdparty.FetchFromAnyProvider && providerID != provider.ID() {
 			continue
 		}
 
-		assetContainer, err := provider.FetchAllAssetsByOwner(ctx, chainID, owner, cursor, limit)
-		if err != nil {
-			log.Error("FetchAllAssetsByOwner failed for", "provider", provider.ID(), "chainID", chainID, "err", err)
-			continue
-		}
+		provider := provider
```

Comment:
🤔

Reply:
ohh, I see, maybe it doesn't nicely handle passing the provider reference to the Functor otherwise. Interesting!